### PR TITLE
Support `torch.concat` alias, add `cat` OpInfo & remove OpInfo test_out  skips {cat, stack, hstack, vtack, dstack}

### LINKED ISSUE #61767: Support `concat` alias to `cat`
## 🚀 Feature
Support `torch.concat`.

## Motivation & Pitch

Python Array API proposes the use of `concat` among other manipulation functions. Currently, PyTorch supports the same functionality with `torch.cat`. As listed in #58742, it would be nice to add an alias `concat` to `cat` making it compliant with Array API.

## Additional context

* [Python Array API concat](https://data-apis.org/array-api/latest/API_specification/manipulation_functions.html#concat-arrays-axis-0)

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

### PR DESCRIPTION
**Pull request by AnirudhDagar**
Fixes #61767

## Changes

- [x] Add `torch.concat` alias to `torch.cat`
- [x] Add OpInfo for `cat`/`concat`
- [x] Fix `test_out` skips (Use `at::native::resize_output` or `at::native::resize_output_check`)
  - [x] `cat`/`concat`
  - [x] `stack`
  - [x] `hstack`
  - [x] `dstack`
  - [x] `vstack`/`row_stack`
- [x] Remove redundant tests for `cat`/`stack` 

~I've not added `cat`/`concat` to OpInfo `op_db` yet, since cat is a little more tricky than other OpInfos (should have a lot of tests) and currently there are no OpInfos for that. I can try to add that in a subsequent PR or maybe here itself, whatever is suggested.~
**Edit**: cat/concat OpInfo has been added.

**Note**: I've added the named tensor support for `concat` alias as well, maybe that's out of spec in `array-api` but it is still useful for consistency in PyTorch.

Thanks to @krshrimali for guidance on my first PR :))

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff @krshrimali

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp
Comment (AnirudhDagar) [inline aten/src/ATen/native/TensorShape.cpp:9]: See line 12 below. This was included twice.

Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:9]: Nice attention to detail

**Comment by mruberry** [comment]
This is cool @AnirudhDagar and the implementation looks correct. 

This does need to an OpInfo for torch.cat and then list torch.concat as an alias of it, however, because otherwise the alias isn't tested. I would just extend this PR with the OpInfo. @krshrimali and @kshitij12345 should be able to help you. I may be unavailable for the rest of the week but if need help merging this PR you can ping @heitorschueroff and @anjali411 and one of them should available to help you.

**Comment by AnirudhDagar** [comment]
Thanks for the review @mruberry! No worries, I'll add the `opinfo` and update the PR. I'm planning to contribute to other missing opinfos later, so this should be a good start.

**Comment by mruberry** [comment]
> Thanks for the review @mruberry! No worries, I'll add the `OpInfo` and update the PR. I'm planning to contribute to other missing opinfos later, so this should be a good start.

Sounds good! It'll be great to have more OpInfos!

**Comment by AnirudhDagar** [comment]
I've added OpInfo for cat/concat. There are a few skips that were required though, namely:

* `SkipInfo('TestCommon', 'test_out')` : I feel this should be fixed as mentioned in the [wiki for out behaviour](https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch). I've followed #42079 for the same. I'd appreciate it if someone could take a look at my [try](url) to fix this skip. In case it is not within the scope of this PR, I'll remove [this commit](https://github.com/pytorch/pytorch/pull/62560/commits/7df250c421aa0d1e916e316c36d09d6c526c1fcd) and keep the `test_out` skips.
* `SkipInfo('TestJit', 'test_variant_consistency_jit')`: JIT Team is working on these, so we can skip those for now.
* `SkipInfo('TestJit', 'test_jit_alias_remapping')`:  The `sample_inputs_cat_concat` returns an iterable of tensors, hence the needed skip. This was discussed in #58838.

File: aten/src/ATen/native/TensorShape.cpp
Comment (anjali411) [inline aten/src/ATen/native/TensorShape.cpp:197]: this might be a good opportunity to switch to structured kernels for `cat`.

cc. @SplitInfinity who might be able to help on the specifics

Comment (AnirudhDagar) [reply aten/src/ATen/native/TensorShape.cpp:197]: I'm not fully aware of the porting to structured kernels, I'm assuming this is the corresponding [RFC](https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md)?

Shouldn't this be left to another PR, or do you feel it is within the scope of this one?

Comment (kshitij12345) [reply aten/src/ATen/native/TensorShape.cpp:197]: If `cat` is not ported to `structured` in this PR, then use `resize_output` from this file. https://github.com/pytorch/pytorch/blob/c7db642a72d80facf9c8a17f42ac0bb04d04723f/aten/src/ATen/native/Resize.cpp#L38-L39

**Review by krshrimali — CHANGES_REQUESTED** [review]
Thanks, @AnirudhDagar for working on this. Looks like a great start, left a few suggestions for sample inputs.

**Comment by AnirudhDagar** [comment]
@krshrimali thanks for reviewing and sharing useful suggestions.
Scalar tensors/arrays (zero-dimensional) are not supported by `torch.{cat,concat}`/`np.concatenate`. Thus, I don't think the test will pass anyway, so we should avoid that.

> It will be good to test when dim is not passed, so something like a dictionary in the cases is useful

That's a good idea, I've updated `numpy` reference implementation to accommodate `axis` instead of `dim` `kwarg`.

I was also able to remove the skip for `test_variant_consistency_jit` with `assert_autodiffed=True`. I've added the comment for `test_jit_alias_remapping ` which is the only skipped test now.

@anjali411 @kshitij12345
I've tweaked `resize_output` to handle `memory_format` which was missing earlier, and now I'm reusing that function in the `cat_out` cpu/cuda kernel.

File: aten/src/ATen/native/Resize.cpp
Comment (kshitij12345) [inline aten/src/ATen/native/Resize.cpp:39]: I don't think we want this behavior, `optional_memory_format` will be applied only if the `output` shape doesn't match with the match shape. (cc: @ngimel for opinion)

Similar discussion: https://github.com/pytorch/pytorch/issues/49878#issuecomment-752191914

File: torch/testing/_internal/common_methods_invocations.py
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:1810]: It would be great if you could concatenate two tensors of different shapes.

**Review by kshitij12345 — COMMENTED** [review]
I think you can remove this test as it should be covered by `test_out` in `test_ops.py`

https://github.com/pytorch/pytorch/blob/3df4870343d212134a92934207bccee91fa6518f/test/test_tensor_creation_ops.py#L711-L754

cc: @mruberry

**Comment by AnirudhDagar** [comment]
@kshitij12345 I've removed the redundant tests you mentioned and some tests from `test_autograd.py` as well. I've updated the OpInfo for the concatenation of two tensors of different shapes.

File: aten/src/ATen/native/Resize.cpp
Comment (mruberry) [reply aten/src/ATen/native/Resize.cpp:39]: @ngimel and I had an opportunity to discuss this offline, here's our thinking:

- memory format is like strides, if a tensor has to be resized then both it and strides can be reset
- if a tensor is the correct shape, however, then strides and memory format should be preserved 

So I don't think we need to pass a memory format argument to resize_output

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:196]: Using resize_output is correct here, to preserve the current behavior of the memory format of the output being derived from the first tensor when a user doesn't specify `out=` then I think the other callers of this function, like here:

https://github.com/pytorch/pytorch/blob/3a88e3fa3176b39dcb5aff9c995b3107a14820ff/aten/src/ATen/native/TensorShape.cpp#L268

would need to change. In particular, for this issue they could allocate an output of the correct size and with the desired memory format in advance. Then resize_output wouldn't resize them. 

This is similar to what structured kernels do, too. cc @bdhirsh to see if we could make cat a structured kernel or not to resolve this issue.

File: aten/src/ATen/native/cuda/Shape.cu
Comment (mruberry) [inline aten/src/ATen/native/cuda/Shape.cu:533]: Same caveat here as before

File: test/test_autograd.py
Comment (mruberry) [inline test/test_autograd.py:2743]: fyi @soulitzer

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:1884]: Nice docstring

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7805]: cc @eellison

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7821]: Awesome skip removals

**Review by mruberry — COMMENTED** [review]
Hey @AnirudhDagar! This looks really good. I made a comment about the out= behavior for cat that I hope is helpful. Let me know if you have any questions. I'll update the Developer FAQ to cover this out= case, too

File: aten/src/ATen/native/TensorShape.cpp
Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:196]: @ngimel and I had an opportunity to review further offline. While dtype is actually computed in advance, it's true that precomputing the shape would be redundant. We also realized we need to further clarify our handling of out= args in the developer wiki.

So here's what we recommend:

- use resize_output_check directly
- if the output needs to be resized, resize it with the memory format of the first tensor (like today)
- if the output doesn't need to be resized, preserve its memory format (don't do anything to it)

This will need a comment to explain why we're not calling resize_output directly

Comment (AnirudhDagar) [reply aten/src/ATen/native/TensorShape.cpp:196]: Thanks, @mruberry @ngimel for reviewing this extensively, especially the resize behaviour in cat_out. I'll address other specific comments in a bit (need a second read to make sure I understand everything correctly). 

IIUC (and I might be jumping the gun here), the final decision looks like the behaviour I had in this [commit](https://github.com/pytorch/pytorch/pull/62560/commits/7df250c421aa0d1e916e316c36d09d6c526c1fcd) previously, but instead of actually writing the warning in here (see how I did it previously in that commit), it's better to use the `resize_output_check` function declared in`Resize.h` which will warn if required.

Comment (mruberry) [reply aten/src/ATen/native/TensorShape.cpp:196]: I believe your understanding is correct; sorry for the confusion and thanks for clarifying

Comment (AnirudhDagar) [reply aten/src/ATen/native/TensorShape.cpp:196]: Awesome! Absolutely no worries, always good to see these discussions (learnt a few new things and also motivates me to look at how the porting to structured kernels is done, maybe sometime later).

 Will go ahead with this approach then. :))

File: aten/src/ATen/native/cuda/Shape.cu
Comment (AnirudhDagar) [reply aten/src/ATen/native/cuda/Shape.cu:533]: Resolved in the conclusion to the discussion above.

File: aten/src/ATen/native/Resize.cpp
Comment (AnirudhDagar) [reply aten/src/ATen/native/Resize.cpp]: Resolved and discussed [here](https://github.com/pytorch/pytorch/pull/62560#discussion_r687363362).

**Comment by AnirudhDagar** [comment]
Thanks for the detailed reviews and for suggesting improvements here. I've updated the PR, resolved conflicts with #63084 and addressed the comments.

Request you all to take a second pass on this. Let me know if it looks good or if we need any changes.

cc @mruberry @ngimel @kshitij12345 @anjali411 @krshrimali

File: test/test_tensor_creation_ops.py
Comment (krshrimali) [inline test/test_tensor_creation_ops.py:754]: I'm not sure if we should remove these tests. If I'm not wrong, OpInfos don't test for cross-device inputs. (I may be wrong :)). cc-ing @kshitij12345

**Review by krshrimali — CHANGES_REQUESTED** [review]
Looks great, thanks @AnirudhDagar. I left a minor comment on removing cross-device tests. I'll be more than happy to approve once that is addressed. :)

File: test/test_tensor_creation_ops.py
Comment (kshitij12345) [reply test/test_tensor_creation_ops.py:754]: https://github.com/pytorch/pytorch/blob/50fc8e8250adafb564e59dc90b39aca6d9b2cee9/test/test_ops.py#L174-L179

https://github.com/pytorch/pytorch/blob/50fc8e8250adafb564e59dc90b39aca6d9b2cee9/test/test_ops.py#L340-L354

IIRC, once `test_out` was added plan was to add OpInfo for `cat` and remove this redundant test. (But we might have to check that all cases are covered). But I am certain that there are few redundant cases.

cc: @mruberry

Comment (krshrimali) [reply test/test_tensor_creation_ops.py:754]: I don't think that the tests you shared check for cases with multiple devices for `inputs`. You are right, there are a few redundant cases, so we can keep those that are not covered by OpInfos.

Example:

```python
torch.cat((cpu_input, cuda_input))
# or
torch.stack((cpu_input, cuda_input))
```

aren't checked.

Comment (AnirudhDagar) [reply test/test_tensor_creation_ops.py:754]: Nice catch @krshrimali! I'll add back the cases which are not covered in OpInfo tests.

Comment (AnirudhDagar) [reply test/test_tensor_creation_ops.py]: @kshitij12345 @krshrimali I've removed the redundant ones and added back the cross device ones in this [commit](https://github.com/pytorch/pytorch/pull/62560/commits/5b54571d2254f59959bbe6beea4ec1a0b39e08aa)

**Comment by AnirudhDagar** [comment]
Hi! I’ve removed the `test_out` skips for the following OpInfos since they were either using `cat_out` itself or were very similar in nature.
- [x] `cat`/`concat`
- [x] `stack`
- [x] `hstack`
- [x] `dstack`
- [x] `vstack`/`row_stack`
 
I thought these skip removals are within the scope of this PR though it was initially meant only for `cat`.

File: aten/src/ATen/native/Resize.h
Comment (mruberry) [inline aten/src/ATen/native/Resize.h:13]: Nice comment extension

File: test/test_tensor_creation_ops.py
Comment (mruberry) [inline test/test_tensor_creation_ops.py:728]: Nice surgical removal of redundant out= testing

File: test/test_tensor_creation_ops.py
Comment (mruberry) [inline test/test_tensor_creation_ops.py:714]: I think this variable and the related test can be removed, too (out= will test for cross-device input x out=)

Comment (mruberry) [reply test/test_tensor_creation_ops.py:714]: There might be a new out= test that needs to be added though to account for cat's memory layout behavior. Is there a test that:

- if out= is the correct shape then the memory format of out= is respected
- if out= is not the correct shape then the memory format is that of the first tensor? 

That's not tested by Opinfo-based out= testing and so needs to be covered specially

**Review by mruberry — COMMENTED** [review]
This looks really good @AnirudhDagar, I have just a couple questions for your review. Looking forward to hearing your thoughts!

File: test/test_tensor_creation_ops.py
Comment (AnirudhDagar) [reply test/test_tensor_creation_ops.py]: > I think this variable and the related test can be removed, too (out= will test for cross-device input x out=)
Done!

**Comment by AnirudhDagar** [comment]
## cat_out memory format behaviour

Replying to [this review comment](https://github.com/pytorch/pytorch/pull/62560#discussion_r690150696) here because it became extremely large while explaining. Apologies for that.

@mruberry this seems more interesting than what I expected initially. The reason is that [`cat_out_cpu`](https://github.com/pytorch/pytorch/blob/9477211e7d609ce382c0e22d7721c14c36d083de/aten/src/ATen/native/TensorShape.cpp#L120) and [`cat_out_cuda`](https://github.com/pytorch/pytorch/blob/9477211e7d609ce382c0e22d7721c14c36d083de/aten/src/ATen/native/cuda/Shape.cu#L454) both follow a slightly different behaviour when resizing an out tensor.

* `cat_out_cpu`: If out= is not the correct shape then the memory format is used from the first tensor.
* `cat_out_cuda`:  If out= is not the correct shape then the memory format is computed using [`compute_output_memory_format `](https://github.com/pytorch/pytorch/blob/9477211e7d609ce382c0e22d7721c14c36d083de/aten/src/ATen/native/cuda/Shape.cu#L435), calling it [here](https://github.com/pytorch/pytorch/blob/9477211e7d609ce382c0e22d7721c14c36d083de/aten/src/ATen/native/cuda/Shape.cu#L512) (**point of difference**) and then that is finally used for reszing. If you look into the implementation it is not directly utilizing the first tensors' memory format unlike the `cpu` variant. To be precise according to [L446 in Shape.cu](https://github.com/pytorch/pytorch/blob/9477211e7d609ce382c0e22d7721c14c36d083de/aten/src/ATen/native/cuda/Shape.cu#L446) `bool contiguous=True` will be assigned even if the first tensor has a different format (eg. `channels_last`) and some other tensor has a `contiguous_format`.

So I wrote a separate test (see below) specifically for testing and making sure that the `cat_out_cuda` fails and passes for different inputs because of it's behaviour. 

First test has inputs with different `memory_format` so it will fail (not behave as mentioned in @mruberry's comment) for `cuda` but pass for `cpu`.
<details>
  <summary>Click to expand: FAILING test_cat_out_memory_format</summary>

  ```python
    def test_cat_out_memory_format(self, device):
        inp_size = (4, 4, 4, 4)
        x_cuda  = torch.randn(inp_size, device=device).contiguous(memory_format=torch.channels_last)
        x_cpu  = torch.randn(inp_size, device='cpu').contiguous(memory_format=torch.channels_last)
        y_cuda = torch.randn(inp_size, device=device)
        y_cpu = torch.randn(inp_size, device='cpu')
  
        # Case 1: if out= is the correct shape then the memory format of out= is respected
        expected_size = (8, 4, 4, 4)
        out_cuda = torch.empty(expected_size, device=device).contiguous(memory_format=torch.contiguous_format)
        res1_cuda = torch.cat((x_cuda, y_cuda), out=out_cuda)
        
        out_cpu = torch.empty(expected_size, device='cpu').contiguous(memory_format=torch.contiguous_format)
        res1_cpu = torch.cat((x_cpu, y_cpu), out=out_cpu)
  
        self.assertTrue(res1_cuda.is_contiguous(memory_format=torch.contiguous_format))
        self.assertTrue(res1_cpu.is_contiguous(memory_format=torch.contiguous_format))
  
        # Case 2: if out= is not the correct shape then the memory format is that of the first tensor
        # output size specifically set wrong so that it is reshaped internally in cat 
        out_cuda = torch.empty((0), device=device).contiguous(memory_format=torch.contiguous_format)
        res2_cuda = torch.cat((x_cuda, y_cuda), out=out_cuda)
  
        out_cpu = torch.empty((0), device='cpu').contiguous(memory_format=torch.contiguous_format)
        res2_cpu = torch.cat((x_cpu, y_cpu), out=out_cpu)
  
        self.assertTrue(res2_cuda.is_contiguous(memory_format=torch.channels_last))
        self.assertTrue(res2_cpu.is_contiguous(memory_format=torch.channels_last))
  ```
</details>

This one has inputs with same `memory_format` but different from out's `memory_format`. The second will pass for both `cpu` and `cuda`.
<details>
  <summary>Click to expand: PASSING test_cat_out_memory_format</summary>

  ```python
    def test_cat_out_memory_format(self, device):
        inp_size = (4, 4, 4, 4)
        x_cuda  = torch.randn(inp_size, device=device).contiguous(memory_format=torch.channels_last)
        x_cpu  = torch.randn(inp_size, device='cpu').contiguous(memory_format=torch.channels_last)
        # Making sure all input tensors follow the same memory_format makes the whole test happy again
        y_cuda = torch.randn(inp_size, device=device).contiguous(memory_format=torch.channels_last)
        y_cpu = torch.randn(inp_size, device='cpu').contiguous(memory_format=torch.channels_last)
  
        # Case 1: if out= is the correct shape then the memory format of out= is respected
        expected_size = (8, 4, 4, 4)
        out_cuda = torch.empty(expected_size, device=device).contiguous(memory_format=torch.contiguous_format)
        res1_cuda = torch.cat((x_cuda, y_cuda), out=out_cuda)
        
        out_cpu = torch.empty(expected_size, device='cpu').contiguous(memory_format=torch.contiguous_format)
        res1_cpu = torch.cat((x_cpu, y_cpu), out=out_cpu)
  
        self.assertTrue(res1_cuda.is_contiguous(memory_format=torch.contiguous_format))
        self.assertTrue(res1_cpu.is_contiguous(memory_format=torch.contiguous_format))
  
        # Case 2: if out= is not the correct shape then the memory format is that of the first tensor
        # output size specifically set wrong so that it is reshaped internally in cat 
        out_cuda = torch.empty((0), device=device).contiguous(memory_format=torch.contiguous_format)
        res2_cuda = torch.cat((x_cuda, y_cuda), out=out_cuda)
  
        out_cpu = torch.empty((0), device='cpu').contiguous(memory_format=torch.contiguous_format)
        res2_cpu = torch.cat((x_cpu, y_cpu), out=out_cpu)
  
        self.assertTrue(res2_cuda.is_contiguous(memory_format=torch.channels_last))
        self.assertTrue(res2_cpu.is_contiguous(memory_format=torch.channels_last))
  ```
</details>

---

### My Opinion:

We should fix this behaviour and make both the cuda and cpu variant behave the same way. We'll need to decide which behaviour we go ahead with. I believe all this (including the tests to be added) should be carried out in a separate issue and a PR because this might be BC Breaking. If you feel the same, I can file an issue and send a subsequent PR for this later.

Ps. I know that `.is_contiguous()` = `.is_contiguous(memory_format=torch.contiguous_format)`. I just went with the latter in the test to be more explicit.

Please let me know if something is unclear.
cc @ngimel @kshitij12345 

---

### Additional Comments

@mruberry @kshitij12345 when digging deeper into memory_format, I [found](https://github.com/pytorch/pytorch/blob/2544664e5470cd5012bd88d07fb573aa31ca8c60/torch/csrc/utils/tensor_memoryformats.cpp#L34) that PyTorch supports three of these in total other than the `torch.preserve_format`:
* `torch.contiguous_format`
* `torch.channels_last`
* `torch.channels_last_3d` (**NOT DOCUMENTED**)

We should add `torch.channels_last_3d` to the [memory format documentation](https://pytorch.org/docs/stable/tensor_attributes.html#torch.torch.memory_format). I completely missed this because it was undocumented.

I can add this to the docs in a separate PR.

File: test/test_autograd.py
Comment (mruberry) [inline test/test_autograd.py:2743]: fyi @soulitzer to review this test -- what's up here?

File: test/test_tensor_creation_ops.py
Comment (mruberry) [inline test/test_tensor_creation_ops.py:718]: Really nice comments

File: docs/source/torch.rst
Comment (mruberry) [inline docs/source/torch.rst:91]: Also needs an entry in tensors.rst

**Review by mruberry — APPROVED** [review]
Nice work, @AnirudhDagar! This was a very challenging PR because it required understanding the new out= behavior, identifying a very subtle discrepancy between cat's CPU and CUDA implementation, and then updating the existing code and tests to account for that.

**Comment by mruberry** [comment]
@krshrimali would you like to take another look?

**Review by krshrimali — APPROVED** [review]
LGTM, great work @AnirudhDagar 🎉 Thanks!

**Pull request closed without merging**
